sdk/identity/src/token_credentials/client_certificate_credentials.rs

```diff
@@ -0,0 +1,232 @@
+use super::{authority_hosts, TokenCredential};
+use azure_core::auth::TokenResponse;
+use base64::{CharacterSet, Config};
+use chrono::Utc;
+use oauth2::AccessToken;
+use openssl::{
+    error::ErrorStack,
+    hash::{hash, DigestBytes, MessageDigest},
+    pkcs12::Pkcs12,
+    pkey::{PKey, Private},
+    sign::Signer,
+    x509::X509,
+};
+use serde::Deserialize;
+use std::{str, time::Duration};
+
+/// Refresh time to use in seconds
+const DEFAULT_REFRESH_TIME: i64 = 300;
+
+/// Base64 encoder for url safe encoding
+const BASE64_URL_SAFE: Config = Config::new(CharacterSet::UrlSafe, false);
+
+/// Provides options to configure how the Identity library makes authentication
+/// requests to Azure Active Directory.
+#[derive(Clone, Debug, PartialEq)]
+pub struct CertificateCredentialOptions {
+    authority_host: String,
+    send_certificate_chain: bool,
+}
+
+impl Default for CertificateCredentialOptions {
+    fn default() -> Self {
+        Self {
+            authority_host: authority_hosts::AZURE_PUBLIC_CLOUD.to_owned(),
+            send_certificate_chain: false,
+        }
+    }
+}
+
+impl CertificateCredentialOptions {
+    /// Create a new TokenCredentialsOptions. default() may also be used.
+    pub fn new(authority_host: String, send_certificate_chain: bool) -> Self {
+        Self {
+            authority_host,
+            send_certificate_chain,
+        }
+    }
+    /// Set the authority host for authentication requests.
+    pub fn set_authority_host(&mut self, authority_host: String) {
+        self.authority_host = authority_host
+    }
+
+    /// The authority host to use for authentication requests.  The default is
+    /// https://login.microsoftonline.com.
+    pub fn authority_host(&self) -> &str {
+        &self.authority_host
+    }
+
+    /// Enable/disable sending the certificate chain
+    pub fn set_send_certificate_chain(&mut self, send_certificate_chain: bool) {
+        self.send_certificate_chain = send_certificate_chain
+    }
+
+    /// Whether certificate chain is sent as part of the request or not. Default is
+    /// set to true
+    pub fn send_certificate_chain(&self) -> bool {
+        self.send_certificate_chain
+    }
+}
+
+/// Enables authentication to Azure Active Directory using a client certificate that
+/// was generated for an App Registration.
+///
+/// In order to use subject name validation send_cert_chain option must be set to true
+/// The certificate is expected to be in base64 encoded der
+pub struct ClientCertificateCredential {
+    tenant_id: String,
+    client_id: String,
+    client_certificate: String,
+    client_certificate_pass: String,
+    options: CertificateCredentialOptions,
+}
+
+impl ClientCertificateCredential {
+    /// Create a new ClientCertificateCredential
+    pub fn new(
+        tenant_id: String,
+        client_id: String,
+        client_certificate: String,
```

Comment:
Docs should underscore that this is the base64 encoded bytes of a cert in PKCS12 format, not a cert in PEM format. Also, to fully support SNI, the credential must accept multiple certs.

Reply:
The api was inspired from the js sdk and there aswell a single certificate is expected. I am not sure if AAD mandates to send the entire chain If there is a change needed i am happy to follow up.

Reply:
AAD requires any intermediate certs in the chain between the signing cert and one signed by a trusted issuer. I expect zero intermediate certs is the common case but if there are any, the client can't authenticate without providing them, so I suggest designing with that in mind.

Re the JS SDK, its certificate credential API was designed before SNI was on the roadmap, and the constructor took a path to a cert file on disk. That didn't preclude supporting SNI because a file can contain multiple certs, at least in PEM format (I'm less familiar with PKCS12). (I'm just trying to provide context here, not suggesting you follow along and take a file path. I think a collection of strings or bytes is better.)

Reply:
@chlowell Thanks for providing context. The chain is supported in the current model. If a chain is needed client_certificate is expected to contain the entire chain. The previous code had the bug where i was also sending the private key and according the docs the private key must not be sent. I have fixed that issue.
For now i would like to resolve this as wont fix and opt for a cleaner, simpler API which aligns with other SDKs. (also i dislike the filepath api so i dont want to add support for that. If somebody else needs it, that can be done in a future iteration)